scripts/calculate_udf_formulas.py

```diff
@@ -0,0 +1,473 @@
+        outp['Volume to take (uL)'] == ng_to_fmol(outp['Amount for prep (ng)'], _outp['Size (bp)']) / nM(inp['Concentration'], inp['Conc. Units'], _outp['Size (bp)'])
+
+        # Calculate from 'Volume to take (uL)'
+        outp['Amount for prep (fmol)'] == outp['Volume to take (uL)'] * nM(inp['Concentration'], inp['Conc. Units'], _outp['Size (bp)'])
+        outp['Amount for prep (ng)'] == outp['Volume to take (uL)'] * ng_ul(inp['Concentration'], inp['Conc. Units'], _outp['Size (bp)'])
+
+
+    3) Prioritized UDFs! Fields signifying the same thing in the calculation may be called
+       different things depending on parent step.
+
+        # Calculate amounts from input amounts by volume fraction
+        outp['Amount (ng)'] = ( outp['Volume to take (uL)'] / inp['Volume (ul)', 'Total Volume (uL)'] ) * inp['Amount (ng)', 'Amount for prep (ng)']
+        outp['Amount (fmol)'] = ( outp['Volume to take (uL)'] / inp['Volume (ul)', 'Total Volume (uL)'] ) * inp['Amount (fmol)', 'Amount for prep (fmol)']
+
+        # Calculate new conc. from new volume
+        outp['Concentration'] = outp['Amount (ng)'] / outp['Volume (ul)']
+        outp['Conc. Units'] = 'ng/ul'
+"""
+
+TIMESTAMP = dt.now().strftime("%y%m%d_%H%M%S")
+
+
+def ng_ul(conc: float, conc_units: str, size: float | None = None) -> float:
+    """Force a concentration to be in ng/ul."""
+    if conc_units == "ng/ul":
+        return conc
+    elif conc_units == "nM":
+        assert size, "Size not provided for conversion to ng"
+        return nM_to_ng_ul(conc, size)
+    else:
+        raise AssertionError(f"Concentration units '{conc_units}' not recognized")
+
+
+def nM(conc: float, conc_units: str, size: float | None = None) -> float:
+    """Force a concentration to be in nM."""
+    if conc_units == "nM":
+        return conc
+    elif conc_units == "ng/ul":
+        assert size, "Size not provided for conversion to nM"
+        return ng_ul_to_nM(conc, size)
+    else:
+        raise AssertionError(f"Concentration units '{conc_units}' not recognized")
+
+
+def assign_val_to_placeholder(
+    val: str | float,
+    placeholder: str,
+    art_in: Artifact | None = None,
+    art_out: Artifact | None = None,
+    step: Process | None = None,
+    overwrite: bool = True,
+) -> None:
+    """Assign a value to a UDF placeholder."""
+    try:
+        udf_name = re.search(r"\['(.*?)'\]", placeholder).groups()[0]  # type: ignore
+    except (AttributeError, IndexError):
+        raise AssertionError(
+            f"Could not extract UDF name from placeholder '{placeholder}'"
+        )
+
+    # Where to assign UDF
+    if "inp" in placeholder:
+        assert art_in, "Input artifact not provided"
+        obj = art_in
+    elif "outp" in placeholder:
+        assert art_out, "Output artifact not provided"
+        obj = art_out
+    elif "step" in placeholder:
+        assert step, "Step not provided"
+        obj = step
+
+    if type(val) is float:
+        val = round(val, 2)
```

Comment:
Couldn't this be an issue sometimes? For display reasons it make sense but if its supposed to be used for calculation I don't think rounding early is a good idea?

Reply:
Very fair, LIMS anyways has it's own way of showing significant digits

d671d2d28329bc724a6b7ed4c8189d4850822d35